### ISSUE TITLE
[sinks] Consider using offset commits for tracking metadata in the Kafka sink

### ISSUE THREAD

**Issue by bkirwi** [MaterializeInc/materialize#14267]
When sinking messages to Kafka, Materialize needs to store and retrieve some metadata to ensure that it publishes messages exactly once to Kafka. Traditionally, it's done this by writing metadata to a "sidecar" topic. This works, but it's a pain to create and configure an additional topic for every sink; we may not have the right permissions, and it's one more thing for users to have to manage and deal with. This is also similar to [how Kafka Connect implements a similar feature](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors).

One proposed alternative is to push the metadata into the headers of the messages we write out. This avoids the sidecar topic, but you no longer have anywhere to store the metadata when you're not publishing messages, and there are some additional risks with compacted topics.

A third option would be to use Kafka's built-in offset storage. The offset store is effectively a `(group, partition) -> (offset, metadata)` key-value store that can be updated as part of a produce transaction. Roughly:
- Instead of adding a message to our current sidecar topic, we'd commit offsets with the necessary timestamps etc. stored in the metadata. (Using something like [`Producer::send_offsets_to_transaction`](https://docs.rs/rdkafka/0.28.0/rdkafka/producer/trait.Producer.html#tymethod.send_offsets_to_transaction).)
- When resuming a failed sink, we'd fetch our committed offsets from Kafka and use that metadata to re-initialize the sink. (Using something like [`Consumer::committed_offsets`](https://docs.rs/rdkafka/0.28.0/rdkafka/consumer/trait.Consumer.html#tymethod.committed_offsets).)

The semantics here are fairly close to the existing sidecar approach, with the internal `__consumer_offsets` topic playing the role of the sidecar topic.

Possible advantages:
- No need for a sidecar topic; less for us and our users to manage.
- Handles the case where we need to mark progress without producing output: just re-commit offsets with updated metadata in their own transaction.

Possible disadvantages:
- It's a bit weird for a producer to commit offsets.
- Materialize will need offset-commit permissions for Kafka, even if it doesn't otherwise consume from Kafka.
- The implementation may be weird with `rdkafka`.
- There's a 4 KiB max on Kafka commit metadata; configurable but annoying to make users change.

Originally discussed [in Slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1660670583791979?thread_ts=1660616551.608819&cid=C01CFKM1QRF).

**Comment by benesch** [comment]
The more I think about this, the more I think that we should just follow what Kafka Connect does. IIUC, that's writing progress messages in our sinks to a separate "offsets" topic. Using consumer group metadata is clever but it's also a bit _weird_, as you note. I think the idea of an offsets topic will be familiar to most folks who are familiar with the Kafka ecosystem; I think a lot of the UX downside of our current sidecar topic is that we conflated the transaction metadata topic with the offsets topic (see discussion in #11216).

Here's my rough thinking. `CREATE SINK ... INTO KAFKA` should take an optional `OFFSETS TOPIC` argument:

```sql
CREATE SINK foo_sink FROM foo INTO KAFKA CONNECTION kafka TOPIC foo (OFFSETS TOPIC my_foo_offsets)
```

If unspecified, the `OFFSETS TOPIC` should default to `{topic-name}_mz_offsets`. 

It should be legal to share offsets topics across multiple Materialize sinks. This should relieve the operational burden for folks who don't want to give us permission to create topics. They can create an offsets topic once, grant us permission, and then set `OFFSETS TOPIC that-topic` for every one of the sinks they create. But I think the right default is to have different offset topics for different sinks by default, as it's not clear how you GC offsets for dropped sinks when they're all mixed together in the same topic. (Idk what Kafka Connect does, but probably you just don't do GC, and hope the topic doesn't get too big.)

The format of the offsets topic should 100% be an implementation detail. We should tell users not to touch or look at the offset topic. I think the schema is quite simple though:

```
key:SINK-ID  value:MZ-TIMESTAMP
```

In short, we just write down the latest timestamp that we've emitted to the sink, keyed by the sink ID. We can encourage users to set a key-based compaction policy on the topic to keep data sizes down.

...I think that's it? Am I missing something?

It occurred to me a few days ago that neither our current sidecar topic nor our proposal of using message headers plays well with compaction. With the old sidecar topic, I don't think we set a message key, so it's not clear what you compact on. And with the message headers proposal, if compaction/retention decides to toss the record with the header, that's a disaster. The design above seems to play really nicely with key-based compaction, because we have a dedicated topic that contains exactly the information we're looking for.

**Comment by aljoscha** [comment]
That seems good, yes. It's also not much of a diff compared to what we have currently. We need to massage SQL syntax and slightly massage what we write into the topic. But that would seem to be it.

**Comment by nmeagan11** [comment]
This looks good to me. I like that the `OFFSETS TOPIC` is optional; my preference is that the user has the least amount to worry about.

**Comment by bkirwi** [comment]
Hey, thanks for the writeup! I agree it seems to be a weirdness vs. extra work tradeoff, and the extra work is both modest and familiar. (And if we change our mind, it seems like a much easier migration to drop the sidecar topic (just stop using it!) than suddenly requiring one.)

> But I think the right default is to have different offset topics for different sinks by default, as it's not clear how you GC offsets for dropped sinks when they're all mixed together in the same topic.

FWIW I think the polite thing is to publish a null value for the old `key: SINK-ID`, which would allow Kafka to compact it out entirely. (And I agree that compaction is a good idea here!)

Assuming we do that, I think I'd be comfortable with recommending a single topic. Less for users to manage, and IIUC the only cost is more records to scan through when resuming a sink... but it sounds like the total size of the topic will never be particularly large?

**Comment by benesch** [comment]
> Hey, thanks for the writeup! I agree it seems to be a weirdness vs. extra work tradeoff, and the extra work is both modest and familiar. (And if we change our mind, it seems like a much easier migration to drop the sidecar topic (just stop using it!) than if we were to suddenly require one for existing sinks.)
> 
> > But I think the right default is to have different offset topics for different sinks by default, as it's not clear how you GC offsets for dropped sinks when they're all mixed together in the same topic.
> 
> FWIW I think the polite thing is to publish a null value for the old `key: SINK-ID`, which would allow Kafka to compact it out entirely. (And I agree that compaction is a good idea here!)

Oh! That's a _very_ good point! Nice.

> Assuming we do that, I think I'd be comfortable with recommending a single topic. Less for users to manage, and IIUC the only cost is more records to scan through when resuming a sink... but it sounds like the total size of the topic will never be particularly large?

Yeah, exactly. It'll be O(n) in the number of sinks. If n is so large that it's prohibitively expensive to read that topic... well I'm sure Materialize has fallen over in a dozen other places trying to manage that many sinks.

So, I'm even more bullish on a single topic for Materialize's offsets. I wonder if we should default to one topic per Kafka _connection_. So every Kafka sink that references the same `CONNECTION` object would use the same offsets topic, unless overridden. We'd name the topic something like `_materialize_offsets_UID`, where `UID` is the ID of the connection object. That'd give us a natural moment to clean up the offsets topic too: when the connection is dropped.

**Comment by nmeagan11** [comment]
Closing in favor of https://github.com/MaterializeInc/materialize/issues/14488.
